src/app/settings/components/project-role-edit-panel/project-role-jwt-tokens.tsx

```diff
@@ -0,0 +1,164 @@
+                        <div className='row'>
+                            <div className='columns small-9'>
+                            <FormField formApi={api} label='' field='expiresIn' component={Text}/>
+                            </div>
+                            <div className='columns small-3'>
+                            <a onClick={() => this.createJWTToken(api, ctx)}>Create token</a>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+                {api.values.token !== '' ? (
+                    <div className='argo-table-list'>
+                    <div className='argo-table-list__head'>
+                        <div className='row'>
+                            <div className='columns small-3'>NEW TOKEN</div>
+                        </div>
+                    </div>
+                    <div className='argo-table-list__row'>
+                        <div className='white-box'><p style={{overflowWrap: 'break-word'}}>{api.values.token}</p></div>
+                    </div>
+                </div>) : null}
+                </form>
+                )}
+                </Form>
+                </React.Fragment>
+            )}
+            </Consumer>
+        );
+    }
+
+    private convertExpiresInToSeconds(expiresIn: string): number {
+        if (!expiresIn) {
+            return 0;
+        }
+        const time = expiresIn.match('^([0-9]+)([smhd])$');
+        const duration = parseInt(time[1], 10);
+        let interval = 1;
+        if (time[2] === 'm') {
+            interval = 60;
+        } else if (time[2] === 'h') {
+            interval = 60 * 60;
+        } else if  (time[2] === 'd') {
+            interval = 60 * 60 * 24;
+        }
+        return duration * interval;
+    }
+
+    private async createJWTToken(api: FormApi, ctx: any) {
+        if (api.errors.expiresIn) {
+            return;
+        }
+        const project = this.props.projName;
+        const role = this.props.roleName;
+        const expiresIn = this.convertExpiresInToSeconds(api.values.expiresIn);
+        let expiresInPrompt = 'has no expiration';
+        if (expiresIn !== undefined || expiresIn !== 0) {
+            expiresInPrompt = 'expires in ' + api.values.expiresIn;
+        }
+        const confirmed = await ctx.popup.confirm(
+            'Create JWT Token', `Are you sure you want to create a JWT token that ${expiresInPrompt} for role '${role}' in project '${project}'?`);
+        if (confirmed) {
+            const jwtToken = await services.projects.createJWTToken({project, role, expiresIn} as CreateJWTTokenParams);
+
+            const updatedProject = await services.projects.get(project);
+            const updatedRole = updatedProject.spec.roles.find((x) => x.name === role);
+            api.setValue('expiresIn', '');
+            api.setValue('token', jwtToken.token);
+            api.setValue('tokens', updatedRole.jwtTokens);
```

Comment:
Reproduced a bug here:

1. enter token expiration and click 'Create' button
2. close panel
3. open panel again

created token disappears from list. Instead of updating form the `ProjectRoleJWTTokensProps` should expose `tokenCreated` callback and parent control should reload role.

Reply:
fixed by moving the creation/deletion functionality up into the project-details component and reloaded the component after submitting the request